bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/scriptengine/InvocationInterceptingScriptEngineWithInvocable.java

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.openhab.automation.jsscripting.internal.scriptengine;
+
+import java.io.Reader;
+
+import javax.script.*;
+
+import org.eclipse.jdt.annotation.NonNull;
+
+/**
+ * Delegate allowing AOP-style interception of calls, either before Invocation, or upon a {@link ScriptException}.
+ * being thrown.
+ *
+ * @author Jonathan Gilbert - Initial contribution
+ *
+ * @param <T> The delegate class
+ */
+public abstract class InvocationInterceptingScriptEngineWithInvocable<T extends ScriptEngine & Invocable>
+        extends DelegatingScriptEngineWithInvocable<T> {
+
+    public InvocationInterceptingScriptEngineWithInvocable(T delegate) {
+        super(delegate);
+    }
+
+    protected void beforeInvocation() {
+    }
+
+    protected @NonNull ScriptException afterThrowsInvocation(@NonNull ScriptException se) {
```

Comment:
Please remove any `@NonNull` annotations, see https://www.openhab.org/docs/developer/guidelines.html#null-annotations. Only `@NonNullByDefault`  and  `@Nullable` should be used.

Reply:
I have a question about this: I originally started with `@NonNullByDefault` but because this is a delegate class which implements something without null annotations, it requires `@NonNull` annotations on pretty much everything in the class.  This is why I inverted the annotations. In this case should I proceed and put `@NonNull` on everything? Or maybe just remove the `@NonNullByDefault` annotation entirely?